File: crates/harness-core/src/llm.rs
use std::pin::Pin;
use std::sync::Arc;

use async_trait::async_trait;
use futures::{stream, Stream};
use serde::{Deserialize, Serialize};

use crate::error::Result;
use crate::memory::{default_estimator, TokenEstimator};
use crate::message::Message;
use crate::tool::ToolSpec;

/// A request to a chat-style LLM. Provider crates translate this into their
/// native wire format.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ChatRequest {
    pub model: String,
    pub messages: Vec<Message>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tools: Vec<ToolSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<u32>,
    /// Provider-issued response id from the previous turn in the same
    /// conversation. The Responses API uses this to reuse server-side
    /// cached state; when paired with `chain_origin` the provider sends
    /// only the *new* messages since the chain anchor and references
    /// the prior response. Other providers (OpenAI Chat Completions,
    /// Anthropic, Google) ignore this field.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub previous_response_id: Option<String>,
    /// Index in `messages` where the chain anchored at
    /// `previous_response_id` begins. The provider sends
    /// `messages[chain_origin..]` as the request input when chaining.
    /// `None` means "send full history" (the historical default).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub chain_origin: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatResponse {
    /// The assistant message returned by the provider. May contain tool_calls.
    pub message: Message,
    pub finish_reason: FinishReason,
    /// Provider-issued response id, when the upstream surface ships
    /// one. Currently populated only by the Responses providers
    /// (OpenAI Responses + Codex) so the caller can chain into the
    /// next request via [`ChatRequest::previous_response_id`].
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub response_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FinishReason {
    Stop,
    ToolCalls,
    Length,
    Other(String),
}

/// Per-call token accounting reported by the provider. Every field is
/// optional because not every endpoint reports every counter — `None`
/// means "the provider didn't tell us", not "zero".
///
/// Providers emit one [`LlmChunk::Usage`] per request whenever the
/// upstream surface exposes counts (OpenAI's `stream_options.include_usage`,
/// Anthropic's `message_delta.usage`, Gemini's `usageMetadata`,
/// Responses' `response.completed.usage`). The agent loop forwards
/// it as [`crate::AgentEvent::Usage`] so transports can show context
/// budget + cache hit savings without each transport re-deriving
/// counts.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Usage {
    /// Input tokens billed for this request, including any cached portion.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prompt_tokens: Option<u32>,
    /// Output tokens billed (assistant text + tool-call args).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub completion_tokens: Option<u32>,
    /// Subset of `prompt_tokens` served from the prompt cache. Useful
    /// for showing the user how much was discount-priced.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cached_prompt_tokens: Option<u32>,
    /// Hidden reasoning tokens billed (Anthropic / Responses reasoning
    /// surfaces). Surfaced separately so callers can attribute cost
    /// without confusing it with completion content.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reasoning_tokens: Option<u32>,
}

/// One piece of a streamed LLM response. Providers emit `ContentDelta` for
/// each partial content token, optionally `ToolCallDelta` for tool-call
/// assembly progress, may emit `Usage` near the end, and finally exactly
/// one `Finish` carrying the fully reconstructed message.
#[derive(Debug, Clone)]
pub enum LlmChunk {
    ContentDelta(String),
    ToolCallDelta {
        index: usize,
        id: Option<String>,
        name: Option<String>,
        arguments_fragment: Option<String>,
    },
    /// Token usage reported by the provider. Optional — most stream
    /// implementations emit zero or one of these, never several.
    Usage(Usage),
    Finish {
        message: Message,
        finish_reason: FinishReason,
        /// Provider-issued identifier for this completion, when the
        /// upstream surface returns one. The Responses API ships an
        /// `id` (e.g. `resp_abc...`) on `response.completed` that can
        /// be threaded into the next request as
        /// `previous_response_id` for server-side state reuse;
        /// non-Responses providers currently leave this `None`.
        #[allow(dead_code)]
        response_id: Option<String>,
    },
}

/// Boxed stream of `LlmChunk` results. The stream ends after the `Finish`
/// variant (or earlier on error).
pub type LlmStream = Pin<Box<dyn Stream<Item = Result<LlmChunk>> + Send>>;

#[async_trait]
pub trait LlmProvider: Send + Sync {
    async fn complete(&self, req: ChatRequest) -> Result<ChatResponse>;

    /// Stream a completion. The default implementation calls `complete` and
    /// emits a single synthesised `Finish` chunk, so providers that do not
    /// yet implement real streaming still satisfy the trait.
    async fn complete_stream(&self, req: ChatRequest) -> Result<LlmStream> {
        let resp = self.complete(req).await?;
        let chunk = LlmChunk::Finish {
            message: resp.message,
            finish_reason: resp.finish_reason,
            response_id: None,
        };
        Ok(Box::pin(stream::once(async move { Ok(chunk) })))
    }

    /// Token-counting strategy for this provider's models. Default: the
    /// `chars / 4 + 4` heuristic in [`crate::CharRatioEstimator`].
    /// Real providers override with a tokeniser-backed implementation
    /// when accuracy matters (memory budgeting, context-window guards).
    fn estimator(&self) -> Arc<dyn TokenEstimator> {
        default_estimator()
    }
}

File: packages/core/src/llm.ts
// LLM provider surface. Ported from harness-core/src/llm.rs.
//
// `Conversation` is the lingua franca; each provider owns conversion to/from
// its native wire format and MUST preserve tool-call/tool-result pairing.
// Streaming lives on its own method (`completeStream`) paralleling `complete`.
import type { Message } from "./message.ts";
import type { ToolSpec } from "./tool.ts";

/**
 * Why a completion ended. Serialises as a snake_case string for the unit
 * cases and as `{ other: "<reason>" }` for anything provider-specific —
 * matching serde's `rename_all = "snake_case"` + `Other(String)`.
 */
export type FinishReason = "stop" | "tool_calls" | "length" | { other: string };

/**
 * Per-call token accounting. Every field is optional: absent means "the
 * provider didn't report this counter", NOT zero.
 */
export interface Usage {
  prompt_tokens?: number;
  completion_tokens?: number;
  cached_prompt_tokens?: number;
  reasoning_tokens?: number;
}

export function emptyUsage(): Usage {
  return {};
}

/** True when no counter was reported. */
export function usageIsEmpty(u: Usage): boolean {
  return (
    u.prompt_tokens === undefined &&
    u.completion_tokens === undefined &&
    u.cached_prompt_tokens === undefined &&
    u.reasoning_tokens === undefined
  );
}

/**
 * Fold `other` into `target` in place. `Some + Some` → saturating sum,
 * `Some + None` keeps existing, `None + Some` adopts new. Mirrors
 * `Usage::add` so the agent loop can roll per-iteration usage into a total.
 */
export function addUsage(target: Usage, other: Usage): void {
  const merge = (a: number | undefined, b: number | undefined): number | undefined => {
    if (a !== undefined && b !== undefined) return a + b;
    if (a === undefined && b !== undefined) return b;
    return a;
  };
  target.prompt_tokens = merge(target.prompt_tokens, other.prompt_tokens);
  target.completion_tokens = merge(target.completion_tokens, other.completion_tokens);
  target.cached_prompt_tokens = merge(target.cached_prompt_tokens, other.cached_prompt_tokens);
  target.reasoning_tokens = merge(target.reasoning_tokens, other.reasoning_tokens);
}

export interface ChatRequest {
  model: string;
  messages: Message[];
  tools?: ToolSpec[];
  temperature?: number;
  max_tokens?: number;
  previous_response_id?: string | null;
  chain_origin?: number | null;
  parallel_tool_calls?: boolean;
}

export interface ChatResponse {
  message: Message;
  finish_reason: FinishReason;
  response_id?: string | null;
  usage?: Usage;
}

/**
 * One piece of a streamed response. Providers emit `content_delta` per token,
 * optionally `tool_call_delta` for argument assembly, may emit one `usage`,
 * and finally exactly one `finish` carrying the reconstructed message.
 *
 * The `usage` chunk's optional `model` names the model that actually produced
 * the tokens. Plain providers leave it unset (the caller falls back to the
 * request model); a routing provider sets it to the routed tier target so token
 * usage is attributed to the model that really ran, not the pre-routing one.
 */
export type LlmChunk =
  | { type: "content_delta"; content: string }
  | {
      type: "tool_call_delta";
      index: number;
      id?: string;
      name?: string;
      arguments_fragment?: string;
    }
  | { type: "usage"; usage: Usage; model?: string }
  | { type: "finish"; message: Message; finish_reason: FinishReason; response_id?: string | null };

export interface LlmProvider {
  complete(req: ChatRequest): Promise<ChatResponse>;
  /** Stream a completion. May return the iterable directly or a Promise of it. */
  completeStream(req: ChatRequest): AsyncIterable<LlmChunk> | Promise<AsyncIterable<LlmChunk>>;
}

/**
 * Default streaming: call `complete` and emit a single synthesised `finish`.
 * Providers without real streaming can delegate to this — mirrors the Rust
 * trait's default `complete_stream` impl.
 */
export async function* defaultCompleteStream(
  provider: Pick<LlmProvider, "complete">,
  req: ChatRequest,
): AsyncGenerator<LlmChunk> {
  const resp = await provider.complete(req);
  yield {
    type: "finish",
    message: resp.message,
    finish_reason: resp.finish_reason,
    response_id: resp.response_id ?? null,
  };
}

/**
 * Optional base class supplying the default `completeStream`. Providers that
 * only implement `complete` can extend this; those with real streaming
 * implement `completeStream` directly against the interface instead.
 */
export class LlmProviderBase implements LlmProvider {
  complete(_req: ChatRequest): Promise<ChatResponse> {
    return Promise.reject(new Error("complete() not implemented"));
  }
  completeStream(req: ChatRequest): AsyncIterable<LlmChunk> {
    return defaultCompleteStream(this, req);
  }
}
